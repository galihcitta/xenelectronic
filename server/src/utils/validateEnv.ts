import { port, cleanEnv } from 'envalid';

export const validateEnv = () => {
  cleanEnv(process.env, {
    PORT: port()
  });
};

// https://www.npmjs.com/package/envalid
