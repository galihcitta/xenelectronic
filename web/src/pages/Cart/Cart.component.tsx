import React from 'react';

import { CartWrapper, Title } from './Cart.styles';

import { CartList } from './cart/CartList.component';
import { Checkout } from './cart/Checkout.component';

export const Cart: React.FC = () => {
  return (
    <>
      <Title>Cart</Title>
      <CartWrapper>
        <CartList />
        <Checkout />
      </CartWrapper>
    </>
  );
};
