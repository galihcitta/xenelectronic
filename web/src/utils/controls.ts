import { Controls } from '../pages/Auth/types';

export const signInControls: Controls = {
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signin-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signin-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};

export const signUpControls: Controls = {
  username: {
    label: 'Enter username',
    placeholder: '',
    type: 'text',
    id: 'signup-username',
    name: 'username',
    value: '',
    validationRules: {
      required: true,
      minLength: 4,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signup-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signup-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true,
      isSecure: true,
      minLength: 8,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};
