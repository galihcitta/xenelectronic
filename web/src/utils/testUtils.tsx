import React from 'react';
import { BrowserRouter, Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../modules/store';
import { createMemoryHistory } from 'history';

const history = createMemoryHistory();

export const withRouter = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <BrowserRouter>
    <Component {...props} />
  </BrowserRouter>
);

export const withHistory = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Router history={history}>
    <Component {...props} />
  </Router>
);

export const withProvider = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Provider store={store}>
    <BrowserRouter>
      <Component {...props} />
    </BrowserRouter>
  </Provider>
);
