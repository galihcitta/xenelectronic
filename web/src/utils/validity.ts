import { ValidationRules } from '../pages/Auth/types';

export const validate = (
  value: string,
  controlName: string,
  rules: ValidationRules
) => {
  let isValid = true;
  let errorMsg = '';

  if (rules.required) {
    isValid = value.trim() !== '' && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} field is required`;
  }

  if (rules.minLength) {
    isValid = value.length >= rules.minLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be at least ${rules.minLength} characters`;
  }

  if (rules.maxLength) {
    isValid = value.length <= rules.maxLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be no longer than ${rules.maxLength} characters`;
  }

  if (rules.isEmail) {
    const pattern = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `email is invalid`;
  }

  if (rules.isSecure) {
    // Minimum eight characters, at least one uppercase letter, one lowercase letter, one number and one special character
    const pattern = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `password too weak`;
  }

  return {
    isValid,
    errorMsg: errorMsg.charAt(0).toUpperCase() + errorMsg.slice(1)
  };
};
