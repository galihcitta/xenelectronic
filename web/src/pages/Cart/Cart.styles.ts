import styled from 'styled-components';

import { device, fontWeight } from '../../utils/styles';

export const CartWrapper = styled.section`
  width: 100%;
  max-width: 500px;
  margin: 0 auto;
`;

export const Title = styled.h2`
  font-size: 1.1rem;
  font-weight: ${fontWeight.bold};
  width: 100%;
  text-align: center;
  margin-bottom: 1em;

  @media ${device.tablet} {
    font-size: 1.4rem;
    margin-bottom: 2em;
  }
`;
