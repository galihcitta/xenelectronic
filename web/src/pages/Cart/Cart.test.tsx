import { render } from '@testing-library/react';
import { Cart } from './Cart.component';
import { withProvider } from '../../utils/testUtils';

const CartWithProvider = withProvider(Cart);

describe('<Cart />', () => {
  it('renders correctly', () => {
    const { asFragment } = render(CartWithProvider({}));
    expect(asFragment()).toMatchSnapshot();
  });
});
