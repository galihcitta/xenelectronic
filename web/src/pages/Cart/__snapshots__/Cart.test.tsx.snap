// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> renders correctly 1`] = `
<DocumentFragment>
  <h2
    class="sc-bwzfXH esHIME"
  >
    Cart
  </h2>
  <section
    class="sc-bdVaJa DuiYN"
  >
    <p
      class="sc-bxivhb cVnBhr"
    >
      Cart is empty.
    </p>
    <div
      class="sc-gqjmRU lkVGqo"
    >
      <p
        class="sc-VigVT giWpjk"
      >
        Total: $ 0
      </p>
      <button
        class="sc-jTzLTM hlRsKz"
        disabled=""
        type="button"
      >
        Checkout
      </button>
      <button
        class="sc-jTzLTM hlRsKz"
        disabled=""
        type="button"
      >
        Clear Cart
      </button>
    </div>
  </section>
</DocumentFragment>
`;
